smartsim/_core/launcher/dragon/dragonConnector.py

```diff
@@ -0,0 +1,447 @@
+
+    @staticmethod
+    def _parse_launched_dragon_server_info_from_iterable(
+        stream: t.Iterable[str], num_dragon_envs: t.Optional[int] = None
+    ) -> t.List[t.Dict[str, str]]:
+        lines = (line.strip() for line in stream)
+        lines = (line for line in lines if line)
+        tokenized = (line.split(maxsplit=1) for line in lines)
+        tokenized = (tokens for tokens in tokenized if len(tokens) > 1)
+        dragon_env_jsons = (
+            config_dict
+            for first, config_dict in tokenized
+            if "DRAGON_SERVER_CONFIG" in first
+        )
+        dragon_envs = [json.loads(config_dict) for config_dict in dragon_env_jsons]
+
+        if num_dragon_envs:
+            sliced_dragon_envs = itertools.islice(dragon_envs, num_dragon_envs)
+            return list(sliced_dragon_envs)
+        return dragon_envs
+
+    @classmethod
+    def _parse_launched_dragon_server_info_from_files(
+        cls,
+        file_paths: t.List[t.Union[str, "os.PathLike[str]"]],
+        num_dragon_envs: t.Optional[int] = None,
+    ) -> t.List[t.Dict[str, str]]:
+        with fileinput.FileInput(file_paths) as ifstream:
+            dragon_envs = cls._parse_launched_dragon_server_info_from_iterable(
+                ifstream, num_dragon_envs
+            )
+
+            return dragon_envs
+
+    @staticmethod
+    def _send_req_with_socket(
+        socket: zmq.Socket[t.Any], request: DragonRequest, flags: int = 0
+    ) -> DragonResponse:
+        client = dragonSockets.as_client(socket)
+        with DRG_LOCK:
+            logger.debug(f"Sending {type(request).__name__}: {request}")
+            client.send(request, flags)
+            response = client.recv()
+
+            logger.debug(f"Received {type(response).__name__}: {response}")
+            return response
+
+
+def _assert_schema_type(obj: object, typ: t.Type[_SchemaT], /) -> _SchemaT:
+    if not isinstance(obj, typ):
+        raise TypeError(f"Expected schema of type `{typ}`, but got {type(obj)}")
+    return obj
+
+
+def _dragon_cleanup(
+    server_socket: t.Optional[zmq.Socket[t.Any]] = None,
+    server_process_pid: t.Optional[int] = 0,
+    server_authenticator: t.Optional[zmq.auth.thread.ThreadAuthenticator] = None,
+    graceful: bool = True,
+) -> None:
+    """Clean up resources used by the launcher.
+    :param server_socket: (optional) Socket used to connect to dragon environment
+    :type server_socket: Optional[zmq.Socket]
+    :param server_process_pid: (optional) Process ID of the dragon entrypoint
+    :type server_process_pid: Optional[int]
+    :param server_authenticator: (optional) Authenticator used to secure sockets
+    :type server_authenticator: Optional[zmq.auth.thread.ThreadAuthenticator]
+    """
+    try:
+        if server_socket is not None:
+            print("Sending shutdown request to dragon environment")
+            # pylint: disable-next=protected-access
+            DragonConnector._send_req_with_socket(
```

Comment:
I'm not sure it makes sense to me to have a module scoped function doing cleanup, but using a static on the class. Can't we just have the class above own cleanup?

I recall a comment about trying to avoid some issue with garbage collection, but again, won't sending a shutdown request satisfy the problem (and maybe if that timesout go nuclear and `os.kill(pid)`

Reply:
I think it is safe now because Dragon has improved its own cleanup.